### PR TITLE
Create notifications backend

### LINKED ISSUE #4: BE: Notifications
Make notifications model. Think about what types of notifications we might want to show the user.
- type
- title
- description
- read/unread
- date
Routes:
- Create notification
- Mark notification as read
- Get all notifications
- Get unread notifications

### PR DESCRIPTION
**Pull request by denjj**
### What this PR does (required):
- Create notifications backend model, route, controller, and validation middleware
- Validation middleware is implemented for POST and PUT requests
- Closes #4 

### Any information needed to test this feature (required):
- With an api testing tool, login as an existing user to retrieve an authorization token
- The following routes can be tested:
  - POST /notifications 
    - Creates a notification for the currently logged in user
    - request body: title, description, type, link
  - PUT /notifications/:notificationId
    - Mark a notification as read
    - request params: notificationId
  - GET /notifications
    - Get all notifications for the logged in user
  - GET /notifications/unread
    - Get all unread notifications for the logged in user

### Any issues with the current functionality (optional):
- Links are currently intended to be implemented as '/dashboard' or '/my-jobs'. There is currently no validation for enforcing this, and no validation to ensure that the link is valid and exists as a route

### CODE REVIEW

**Comment by denjj** [comment]
I'm merging this PR before approval to have our project ready for the presentation

File: server/routes/notification.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  markNotificationAsRead,
+  getAllNotifications,
+  getUnreadNotifications,
+} = require("../controllers/notification");
+const {
+  validateCreateNotification,
+  validateMarkNotificationAsRead,
+} = require("../middleware/notification.validate");
+
+router.route("/").post(protect, validateCreateNotification, createNotification);
```
Comment (rajivtitus) [inline server/routes/notification.js:15]: Nice routes and validation middleware 👍🏽

File: server/models/Notification.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    required: true,
+  },
+  status: {
+    type: String,
+    enum: ["read", "unread"],
+    default: "unread",
+    lowercase: true,
+    required: true,
+  },
+  date: {
```
Comment (rajivtitus) [inline server/models/Notification.js:39]: [Mongoose schemas have a timestamps option](https://mongoosejs.com/docs/guide.html#timestamps) that tells Mongoose to automatically manage createdAt and updatedAt properties on your documents

File: server/models/Notification.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    enum: ["booking", "payment", "system"],
+    lowercase: true,
+    required: true,
+  },
+  link: {
+    type: String,
+    lowercase: true,
+    required: true,
+  },
+  status: {
```
Comment (rajivtitus) [inline server/models/Notification.js:32]: Renaming this field to `read` and making it a boolean type seems more appropriate here

File: server/models/Notification.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  },
+  description: {
+    type: String,
+    required: true,
+  },
+  photo: {
+    type: String,
+    default: "",
+  },
+  type: {
```
Comment (rajivtitus) [inline server/models/Notification.js:21]: 👍🏽

File: server/models/Notification.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+const mongoose = require("mongoose");
+
+const notificationSchema = new mongoose.Schema({
+  user: {
```
Comment (rajivtitus) [inline server/models/Notification.js:4]: This field keeps track of the recipient. How do we keep track of the sender/creator?

File: server/controllers/notification.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    status: "unread",
+  })
+    .populate({
+      path: "user",
+      model: "user",
+      select: "name email",
+    })
+    .sort({ date: "desc" });
+
+  if (!notifications) {
```
Comment (rajivtitus) [inline server/controllers/notification.js:102]: This 'if' clause can be removed as a mongoose `find()` query returns an empty list when no matches are found. Ideally we want to return this list to the client to have a consistent response format

File: server/controllers/notification.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    userId: req.user.id,
+  })
+    .populate({
+      path: "user",
+      model: "user",
+      select: "name email",
+    })
+    .sort({ date: "desc" });
+
+  if (!notifications) {
```
Comment (rajivtitus) [inline server/controllers/notification.js:72]: Same as comment below!

File: server/controllers/notification.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  const { notificationId } = req.params;
+
+  const notification = await Notification.findById(notificationId);
+
+  if (!notification) {
+    res.status(404);
+    throw new Error("Notification not found");
+  }
+
+  if (notification.user.toString() !== req.user.id) {
```
Comment (rajivtitus) [inline server/controllers/notification.js:41]: 💯

**Review by rajivtitus — COMMENTED** [review]
Impressive work with this! Just a few comments for you to have a look and address :)

**Pull request merged**